src/gradient8.c

```diff
@@ -0,0 +1,96 @@
+#define TOPOTOOLBOX_BUILD
+
+#include <math.h>
+#include <omp.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "topotoolbox.h"
+
+#define PI 3.14159265358979323846f
+#define SQRT2 1.41421356237309504880f
+/*
+  Compute the gradient of each cell in a digital elevation model.
+
+  The arrays pointed to by `gradient` and `dem` should represent
+  two-dimensional arrays of size specified in dims. The gradient is
+  computed as the maximum difference in elevation between a cell
+  and its 8 neighbors.
+
+  int use_mp:
+    0   --> don't use multiprocessing
+    1   --> use multiprocessing
+
+  char unit:
+    't' --> tangent (default)
+    'r' --> radian
+    'd' --> degree
+    's' --> sine
+    'p' --> percent
+ */
+
+TOPOTOOLBOX_API
+void gradient8(float *output, float *dem, float cellsize, char unit, int use_mp,
+               ptrdiff_t dims[2]) {
+  ptrdiff_t i_offset[8] = {-1, -1, -1, 0, 0, 1, 1, 1};
+  ptrdiff_t j_offset[8] = {-1, 0, 1, -1, 1, -1, 0, 1};
+
+#ifdef _MSC_VER
+  ptrdiff_t j;
+#pragma omp parallel for if (use_mp)
+  for (j = 0; j < dims[1]; j++) {
+    for (ptrdiff_t i = 0; i < dims[0]; i++) {
+#else
+  ptrdiff_t i, j;
+#pragma omp parallel for collapse(2) if (use_mp)
+  for (j = 0; j < dims[1]; j++) {
+    for (i = 0; i < dims[0]; i++) {
+#endif
```

Comment:
I agree it look pretty ugly. For our case, where the inner loop is always the same operation, it seems like there is not huge performance gain to be had when using the `collapse` function. So I think we would be fine to remove it for gradient8. I will still try to implement it with the flag so that we have an example for later functions.